File: Javascript/merge_sort.js
var mergeSort = function(array) {
  if (array.length === 1) {
    return array;
  }
  var midpoint = array.length/2;
  var leftArray = mergeSort(array.slice(0, midpoint));
  var rightArray = mergeSort(array.slice(midpoint, array.length));
  return merge(leftArray, rightArray);
};

var merge = function(leftArray, rightArray) {
  var mergedArray = [];
  var leftIndex = 0;
  var rightIndex = 0;
  while (leftIndex < leftArray.length && rightIndex < rightArray.length) {
    if (leftArray[leftIndex] < rightArray[rightIndex]) {
      mergedArray.push(leftArray[leftIndex]);
      leftIndex += 1;
    } else if (leftArray[leftIndex] > rightArray[rightIndex]) {
      mergedArray.push(rightArray[rightIndex]);
      rightIndex += 1;
    } else {
      mergedArray.push(leftArray[leftIndex], rightArray[rightIndex]);
      leftIndex += 1;
      rightIndex += 1;
    }
  }
  if (leftIndex < leftArray.length) {
    Array.prototype.push.apply(mergedArray, leftArray.slice(leftIndex, leftArray.length));
  } else if (rightIndex < rightArray.length) {
    Array.prototype.push.apply(mergedArray, rightArray.slice(rightIndex, rightArray.length));
  }
  return mergedArray;
}

File: Java/src/MergeSort.java
import java.util.ArrayList;
import java.util.Arrays;

public class MergeSort {

	public static int[] mergeSort(int[] arr) {
		if (arr.length <= 1) {
			return arr;
		} else {
			int midpoint = arr.length/2;
			int[] leftArr = mergeSort(Arrays.copyOfRange(arr, 0, midpoint));
			int[] rightArr = mergeSort(Arrays.copyOfRange(arr, midpoint, arr.length));
			return merge(leftArr, rightArr);
		}
	}
	
	public static int[] merge(int[] leftArr, int[] rightArr) {
		int leftIndex = 0;
		int rightIndex = 0;
		ArrayList<Integer> mergedArray = new ArrayList<Integer>();
		
		while (leftIndex < leftArr.length && rightIndex < rightArr.length) {
			if (leftArr[leftIndex] <= rightArr[rightIndex]) {
				mergedArray.add(leftArr[leftIndex]);
				leftIndex += 1;
			} else {
				mergedArray.add(rightArr[rightIndex]);
				rightIndex += 1;
			}
		}
		
		if (leftIndex < leftArr.length) {
			while (leftIndex < leftArr.length) {
				mergedArray.add(leftArr[leftIndex]);
				leftIndex += 1;
			}
		} else if (rightIndex < rightArr.length) {
			while (rightIndex < rightArr.length) {
				mergedArray.add(rightArr[rightIndex]);
				rightIndex += 1;
			}
		}
		
		int[] mergedIntArray = mergedArray.stream().mapToInt(i -> i).toArray();
		return mergedIntArray;
	}
	
	public static void main(String[] args) {
		int[] arr = {5, 3, 7, 98, 2, 43, 2, 5, 6};
		System.out.println("Unsorted array:");
		System.out.println(Arrays.toString(arr));
		System.out.println("Sorted array:");
		System.out.println(Arrays.toString(mergeSort(arr)));
	}

}
